lib/chat_models/chat_google_ai.ex

```diff
@@ -0,0 +1,432 @@
+        {:error, "Request timed out"}
+
+      other ->
+        Logger.error("Unexpected and unhandled API response! #{inspect(other)}")
+        other
+    end
+  end
+
+  def do_api_request(%ChatGoogleAI{stream: true} = google_ai, messages, functions, callback_fn) do
+    Req.new(
+      url: build_url(google_ai),
+      json: for_api(google_ai, messages, functions),
+      receive_timeout: google_ai.receive_timeout,
+      finch_request:
+        Utils.finch_stream_fn(google_ai, &do_process_response(&1, MessageDelta), callback_fn)
+    )
+    |> Req.Request.put_header("accept-encoding", "utf-8")
+    |> Req.post()
+    |> case do
+      {:ok, %Req.Response{body: data}} ->
+        # Google AI uses `finishReason: "STOP` for all messages in the stream.
+        # This field can't be used to terminate the list of deltas, so simulate
+        # this behavior by forcing the final delta to have `status: :complete`.
+        complete_final_delta(data)
+
+      {:error, %LangChainError{message: reason}} ->
+        {:error, reason}
+
+      other ->
+        Logger.error(
+          "Unhandled and unexpected response from streamed post call. #{inspect(other)}"
+        )
+
+        {:error, "Unexpected response"}
+    end
+  end
+
+  @spec build_url(t()) :: String.t()
+  defp build_url(%ChatGoogleAI{endpoint: endpoint, version: version, model: model} = google_ai) do
+    "#{endpoint}/#{version}/models/#{model}:#{get_action(google_ai)}?key=#{get_api_key(google_ai)}"
+    |> use_sse(google_ai)
+  end
+
+  @spec use_sse(String.t(), t()) :: String.t()
+  defp use_sse(url, %ChatGoogleAI{stream: true}), do: url <> "&alt=sse"
+  defp use_sse(url, _model), do: url
+
+  @spec get_action(t()) :: String.t()
+  defp get_action(%ChatGoogleAI{stream: false}), do: "generateContent"
+  defp get_action(%ChatGoogleAI{stream: true}), do: "streamGenerateContent"
+
+  def complete_final_delta(data) when is_list(data) do
+    update_in(data, [Access.at(-1), Access.at(-1)], &%{&1 | status: :complete})
+  end
+
+  def do_process_response(response, message_type \\ Message)
+
+  def do_process_response(%{"candidates" => candidates}, message_type) when is_list(candidates) do
+    candidates
+    |> Enum.map(&do_process_response(&1, message_type))
+  end
+
+  def do_process_response(
+        %{
+          "content" => %{"parts" => [%{"functionCall" => %{"args" => raw_args, "name" => name}}]}
+        } = data,
+        message_type
+      ) do
+    case message_type.new(%{
+           "role" => "assistant",
+           "function_name" => name,
+           "arguments" => raw_args,
+           "complete" => true,
+           "index" => data["index"]
+         }) do
+      {:ok, message} ->
+        message
+
+      {:error, changeset} ->
+        {:error, Utils.changeset_error_to_string(changeset)}
+    end
+  end
+
+  def do_process_response(
+        %{
+          "finishReason" => finish,
+          "content" => %{"parts" => parts, "role" => role},
```

Comment:
parts contains the actual text segments returned by the LLM. It's normally only a single part actually, but i think if you are using their multi-modal model i.e. `gemini-pro-vision`, it can return both text and image parts in the response. I haven't tried any of the image stuff though. My guess is that the API response format is just future proofed so they can add additional capabilities without needing to break their contracts.